### ISSUE TITLE
Failed to extract docs from crate

### ISSUE THREAD

**Issue by whatakey** [kellnr/kellnr#174]
Pressing the build button in documentation from the WebUI gives the following error:

![Untitled](https://github.com/kellnr/kellnr/assets/126112966/e94cb3b2-8041-4934-b5d0-187e39965362)

```
ERROR docs::doc_queue: Failed to extract docs from crate: could not execute process `rustc -vV` (never executed)
```

I have full rust stable toolchain installed:
```
$ rustc -vV
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: aarch64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```
```
$ rustdoc -V
rustdoc 1.75.0 (82e1608df 2023-12-21)
```

**Comment by secana** [comment]
I tried to replicate the issue, but failed. Do have more information about your setup, so I can replicate it?

**Comment by whatakey** [comment]
Hello,

I have kellnr v5.1.1 manually installed on RPI4 (aarch64) running Arch Linux

I see folders get generated in the dir "doc_queue" when I press build button, but nothing is in docs dir (always empty)

![Untitled](https://github.com/kellnr/kellnr/assets/126112966/d4bdcc24-1ae9-49e3-8ce2-81f26fcf67c3)

**Comment by CvH** [comment]
got the same error at ubuntu 22.04 installed kellnr as binary, get startet from user kellnr
rust is installed at that user

```
root@rust-crate-registry:~# systemctl status kellnr.service 
* kellnr.service - Kellnr Crate Registry
     Loaded: loaded (/etc/systemd/system/kellnr.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2024-01-26 13:47:16 UTC; 53s ago
   Main PID: 110 (kellnr)
      Tasks: 8 (limit: 154316)
     Memory: 39.6M
        CPU: 54ms
     CGroup: /system.slice/kellnr.service
             `-110 /kellnr/kellnr

Jan 26 13:47:16 rust-crate-registry systemd[1]: Started Kellnr Crate Registry.
Jan 26 13:47:16 rust-crate-registry kellnr[110]: 2024-01-26T13:47:16.575371Z  INFO kellnr: Starting kellnr
Jan 26 13:48:06 rust-crate-registry kellnr[110]: 2024-01-26T13:48:06.835355Z ERROR docs::doc_queue: Failed to extract docs from crate: could not execute process `rustc -vV` (never executed)
```

```
kellnr@rust-crate-registry:~$ rustc -vV
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```

**Comment by secana** [comment]
I installed kellnr on a fresh Ubuntu 22.04 aarch64 and tested the doc creation. As it works out of the box for me, could you provide the logs from kellnr with the log_level "trace" when you try to build docs?

**Comment by CvH** [comment]
I found the problem after you stated it just works. 
I had installed rust via https://www.rust-lang.org/ and not via apt.
After deinstalling rust and installing it with `apt install rust-all` it worked like a charm 👍 

```
kellnr@rust-crate-registry:~$ rustc -vV
rustc 1.71.1 (eb26296b5 2023-08-03) (built from a source tarball)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

No idea if this is a bug or a feature.

**Comment by whatakey** [comment]
@secana thank you for testing.

unfortunately now for some reason the build button disappeared entirely to an unknown destination:

![Untitled](https://github.com/kellnr/kellnr/assets/126112966/d3cc3285-2a45-46c0-95d9-a853de2a14ab)

and my poor little crate is stuck in the doc queue for few days now:

![2](https://github.com/kellnr/kellnr/assets/126112966/b03c9d2f-37a6-4714-81a5-7786eda88b17)

I guess logs aren't much of use because I can't press (build) docs but here are the lines just after a kellnr service restart:

```
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.334757Z TRACE index::cratesio_prefetch_api: Index not-modified for hermit-abi
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.334872Z TRACE index::cratesio_prefetch_api: Updating prefetch data for gimli
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.363544Z TRACE index::cratesio_prefetch_api: Index not-modified for siphasher
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.398384Z TRACE index::cratesio_prefetch_api: Index not-modified for range-alloc
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.552497Z TRACE index::cratesio_prefetch_api: Index not-modified for tinyvec_macros
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.570490Z TRACE index::cratesio_prefetch_api: Index not-modified for gpu-allocator
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.598523Z TRACE index::cratesio_prefetch_api: Index not-modified for bit-vec
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.608568Z TRACE index::cratesio_prefetch_api: Index not-modified for glow
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.638191Z TRACE index::cratesio_prefetch_api: Index not-modified for com-rs
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.671360Z TRACE index::cratesio_prefetch_api: Index not-modified for termcolor
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.696598Z TRACE index::cratesio_prefetch_api: Index not-modified for backtrace
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.778934Z TRACE index::cratesio_prefetch_api: Index not-modified for widestring
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.788314Z TRACE index::cratesio_prefetch_api: Index not-modified for gpu-descriptor-types
Feb 01 22:54:15 OnlyACs kellnr[54990]: 2024-02-01T21:54:15.912848Z TRACE index::cratesio_prefetch_api: Index not-modified for objc_exception
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.074274Z TRACE index::cratesio_prefetch_api: Index not-modified for zerocopy
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.076409Z TRACE index::cratesio_prefetch_api: Index not-modified for addr2line
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.088106Z TRACE index::cratesio_prefetch_api: Index not-modified for object
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.107444Z TRACE index::cratesio_prefetch_api: Index not-modified for rustc-demangle
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.111063Z TRACE index::cratesio_prefetch_api: Index not-modified for gpu-alloc-types
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.122312Z TRACE index::cratesio_prefetch_api: Index not-modified for zerocopy-derive
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.148210Z TRACE index::cratesio_prefetch_api: Index not-modified for unicode-width
Feb 01 22:54:16 OnlyACs kellnr[54990]: 2024-02-01T21:54:16.198190Z TRACE index::cratesio_prefetch_api: Index not-modified for gimli
Feb 01 22:54:30 OnlyACs kellnr[54990]: 2024-02-01T21:54:30.582659Z DEBUG request: tower_http::trace::on_request: started processing request method=GET uri=/api/v1/ui/statistic version=HTTP/1.1
Feb 01 22:54:30 OnlyACs kellnr[54990]: 2024-02-01T21:54:30.595975Z DEBUG request: tower_http::trace::on_response: finished processing request latency=13 ms status=200 method=GET uri=/api/v1/ui/statistic version=HTTP/1.1
Feb 01 22:54:41 OnlyACs kellnr[54990]: 2024-02-01T21:54:41.179925Z DEBUG request: tower_http::trace::on_request: started processing request method=GET uri=/api/v1/ui/crates?page=0&page_size=12 version=HTTP/1.1
Feb 01 22:54:41 OnlyACs kellnr[54990]: 2024-02-01T21:54:41.181777Z DEBUG request: tower_http::trace::on_response: finished processing request latency=1 ms status=200 method=GET uri=/api/v1/ui/crates?page=0&page_size=12 version=HTTP/1.1
```

I'll re-install kellnr later and see if the build docs comes back.

**Comment by secana** [comment]
@whatakey how did you install rust? With rustup or from the distributions package manager?

**Comment by whatakey** [comment]
@secana
I installed Rust using rustup

**Comment by CvH** [comment]
> I cannot reproduce the issue

It may be helpful to add this to the docs, its not really obvious that the apt install behaves different to the rustup ones.
So apt install rust works, rustup not (how they get installed differs a lot).

**Comment by SmnTin** [comment]
I ran into the same problem. The docs don't build if `kellnr` is run as a `systemd` service and Rust is installed via `rustup`.

The problem is that `rustup` installs Rust locally in the home directory and adds a line to `.bashrc` to modify `$PATH`. When `systemd` starts a service, it doesn't source `.bashrc`. So, when `kellnr` tries to build docs, it cannot find the Rust installation because it is not on `$PATH` for that process.

This is why installing Rust via `apt` works but via `rustup` doesn't. `apt` installs in `/usr/bin` (or similar) which is always on `$PATH`.

The solution is to manually specify the correct `$PATH` in the `systemd` service definition:
```ini
# ...

[Service]
Environment=PATH=/path/to/home/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
# ...

# ...
```

**Comment by sorenrade** [comment]
Running on Ubuntu 24.04 installed Rust via rustup as my non-root user. 

Also had this issue. @SmnTin 's solution helped, but I also needed to set `Service.User` to my non-root user. Then since I wanted to open port 80 directly used the + in ExecStart [(docs here)](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Command%20lines) to re-elevate my privileges. I don't fully understand the security implications of +, so use that bit as your own risk. 



```
[Service]
User=<MY NON ROOT USER THAT INSTALLED RUSTUP>
ExecStart=+/bin/bash -c "PATH=/home/<SAME USER>/.cargo/bin:$PATH <ABSOLUE PATH TO KELLNR>"
Restart=always
RestartSec=1

``` 

This only fixes generating the docs for the top-level crate. I think it's also possible that you get the error if the registry is misconfigured in the .cargo/config.toml on your server. Mine was a LAN address, for example, so I got the error again when generating docs for a private crate referenced by another crate. I had to use the public address of my server to fully stop the errors.
